packages/eslint-plugin/tests/rules/no-array-delete.test.ts

```diff
@@ -0,0 +1,519 @@
+          line: 3,
+          column: 9,
+          endColumn: 29,
+          suggestions: [
+            {
+              messageId: 'useSplice',
+              data: {
+                key: '0',
+                target: 'getArray()',
+              },
+              output: `
+        declare function getArray<T extends number[]>(): T;
+        getArray().splice(0, 1);
+      `,
+            },
+          ],
+        },
+      ],
+    },
+
+    {
+      code: `
+        declare function getArray<T extends number>(): T[];
+        delete getArray()[0];
+      `,
+      errors: [
+        {
+          messageId: 'noArrayDelete',
+          line: 3,
+          column: 9,
+          endColumn: 29,
+          suggestions: [
+            {
+              messageId: 'useSplice',
+              data: {
+                key: '0',
+                target: 'getArray()',
+              },
+              output: `
+        declare function getArray<T extends number>(): T[];
+        getArray().splice(0, 1);
+      `,
+            },
+          ],
+        },
+      ],
+    },
+
+    {
+      code: `
+        function deleteFromArray(a: number[]) {
+          delete a[0];
+        }
+      `,
+      errors: [
+        {
+          messageId: 'noArrayDelete',
+          line: 3,
+          column: 11,
+          endColumn: 22,
+          suggestions: [
+            {
+              messageId: 'useSplice',
+              data: {
+                key: '0',
+                target: 'a',
+              },
+              output: `
+        function deleteFromArray(a: number[]) {
+          a.splice(0, 1);
+        }
+      `,
+            },
+          ],
+        },
+      ],
+    },
+
+    {
+      code: `
+        function deleteFromArray<T extends number>(a: T[]) {
+          delete a[0];
+        }
+      `,
+      errors: [
+        {
+          messageId: 'noArrayDelete',
+          line: 3,
+          column: 11,
+          endColumn: 22,
+          suggestions: [
+            {
+              messageId: 'useSplice',
+              data: {
+                key: '0',
+                target: 'a',
+              },
+              output: `
+        function deleteFromArray<T extends number>(a: T[]) {
+          a.splice(0, 1);
+        }
+      `,
+            },
+          ],
+        },
+      ],
+    },
+
+    {
+      code: `
+        function deleteFromArray<T extends number[]>(a: T) {
+          delete a[0];
+        }
+      `,
+      errors: [
+        {
+          messageId: 'noArrayDelete',
+          line: 3,
+          column: 11,
+          endColumn: 22,
+          suggestions: [
+            {
+              messageId: 'useSplice',
+              data: {
+                key: '0',
+                target: 'a',
+              },
+              output: `
+        function deleteFromArray<T extends number[]>(a: T) {
+          a.splice(0, 1);
+        }
+      `,
+            },
+          ],
+        },
+      ],
+    },
```

Comment:
This test is failing and I have no idea why. Help would be greatly appreciated

Reply:
✨ `getConstrainedTypeAtLocation` ✨ 

Type parameters are funky because they're not equal to the type they're constrained to. We just conceptually think of them as that type. This trips up rules a lot 😄

Reply:
Thank you!